pkg/cmd/pulumi/templates/templates.go

```diff
@@ -0,0 +1,211 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package templates adds an abstraction for project templates that may be local or
+// remote.
+//
+// All templates are convertible into [workspace.Template].
+package templates
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+// Source provides access to a set of project templates, any set of which may be present on
+// disk.
+//
+// Source is responsible for cleaning up old templates, and should always be [Close]d when
+// created.
+type Source struct {
+	templates    []Template
+	errorOnEmpty []error
+	errors       []error
+	// closers holds a list of functions to be invoked when the Source is closed.
+	//
+	// closers always has length at least one, and the first closer always cancels the
+	// context associated with fetch operations.
+	closers []func() error
+	closed  bool
+
+	// m should be held whenever Source is mutated.
+	m  sync.Mutex
+	wg sync.WaitGroup
+}
+
+// Templates lists the templates available to the [Source].
+//
+// Templates *does not* produce a sorted list. If templates need to be sorted, then the
+// caller is responsible for sorting them.
+func (s *Source) Templates() ([]Template, error) {
+	s.wg.Wait() // Wait to ensure that all templates have been fetched before returning the template list.
+
+	s.lockOpen("read templates")
+	defer s.m.Unlock()
+	if err := errors.Join(s.errors...); err != nil {
+		return nil, err
+	}
+	if len(s.templates) == 0 {
+		return nil, errors.Join(s.errorOnEmpty...)
+	}
+	return s.templates, nil
+}
+
+func (s *Source) addTemplate(t Template) {
+	contract.Assertf(t != nil, "We should never return nil templates")
+	s.lockOpen("add template")
+	s.templates = append(s.templates, t)
+	s.m.Unlock()
+}
+
+func (s *Source) addCloser(f func() error) {
+	s.lockOpen("add closer")
+	s.closers = append(s.closers, f)
+	s.m.Unlock()
+}
+
+func (s *Source) addError(err error) {
+	s.lockOpen("add error")
+	s.errors = append(s.errors, err)
+	s.m.Unlock()
+}
+
+func (s *Source) addErrorOnEmpty(err error) {
+	s.lockOpen("add error on empty")
+	s.errorOnEmpty = append(s.errorOnEmpty, err)
+	s.m.Unlock()
+}
+
+func (s *Source) lockOpen(action string) {
+	s.m.Lock()
+	contract.Assertf(!s.closed, "Attempted to act on closed source: "+action)
+}
+
+// Close cleans up the [Source] and any associated templates.
+//
+// Close should always be called when [Source] is dropped.
+func (s *Source) Close() error {
+	contract.IgnoreError(s.closers[0]())
```

Comment:
Probably deserves a comment that we're calling the "cancel" closer here.  But also maybe we should have that separate rather than in the list of other "closers", since there can only ever be one, and it's kinda unrelated to the others.